Windows: ship with common tools

justinmk opened the pull request:
#1811

equalsraf committed:
win/build: Download winpty

Winpty has x86/x64 binary builds, download them when building
Neovim.

equalsraf committed:
win/build: Fix libuv recipe for Windows/NMake X86_64

The 64bit check for the libuv recipe worked for the VS generator
but not for NMake.

equalsraf committed:
win/build: detect architecture

justinmk committed:
win/package: provide common tools

References #1507
Closes #1811

curl.exe curl_7_52_1_openssl_nghttp2_x86.7z from https://winampplugins.co.uk/curl/
    curl 7.52.1 (x86_64-pc-win32) libcurl/7.52.1 OpenSSL/1.0.2k zlib/1.2.8 nghttp2/1.19.0
    Protocols: dict file ftp ftps gopher http https imap imaps ldap pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
    Features: AsynchDNS IPv6 Largefile NTLM SSL libz HTTP2 HTTPS-proxy

tidy.exe http://tidybatchfiles.info
    HTML Tidy for Windows
    tidy version 5.3.14 date 2017.01.09 compiled for Windows win32 x86.

7za.exe http://www.7-zip.org
    7-Zip 16.04 (2016-10-04)

cat.exe http://unxutils.sourceforge.net

from gVim:
    diff.exe GNU diffutils version 2.7
    xxd.exe V1.10 27oct98 by Juergen Weigert (Win32)
    ye olde hacked-up tee.exe

justinmk commented:
```
cd /d C:\projects\neovim\build\src\nvim && C:\msys64\mingw64\bin\cmake.exe -E make_directory C:/projects/neovim/build/windows_runtime_deps
cd /d C:\projects\neovim\build\src\nvim && C:\msys64\mingw64\bin\cmake.exe -DCMAKE_PREFIX_PATH=C:/projects/neovim/.deps/usr -DBINARY="C:/projects/neovim/.deps/usr/bin/swiss.exe" -DDST=C:/projects/neovim/build/windows_runtime_deps -P C:/projects/neovim/cmake/WindowsDllCopy.cmake
CMake Error at C:/msys64/mingw64/share/cmake-3.4/Modules/GetPrerequisites.cmake:798 (message):
  C:/Program Files (x86)/Microsoft Visual Studio 9.0/VC/bin/dumpbin.exe
  failed: 1000
  
  Microsoft (R) COFF/PE Dumper Version 9.00.30729.01
  Copyright (C) Microsoft Corporation.  All rights reserved.
  
  
  Dump of file C:/projects/neovim/.deps/usr/bin/swiss.exe
  
  File Type: EXECUTABLE IMAGE
  
  LINK : fatal error LNK1000: Internal error during DumpSections
  
    Version 9.00.30729.01
  
    ExceptionCode            = C0000005
    ExceptionFlags           = 00000000
    ExceptionAddress         = 009771A0 (00950000) "C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\bin\link.exe"
    NumberParameters         = 00000002
    ExceptionInformation[ 0] = 00000000
    ExceptionInformation[ 1] = 00000004
  
  CONTEXT:
    Eax    = 40000040  Esp    = 001BDEF0
    Ebx    = 028B3918  Ebp    = 001BDF18
    Ecx    = 00000004  Esi    = 00000004
    Edx    = 0095405C  Edi    = 0000014C
    Eip    = 009771A0  EFlags = 00010246
    SegCs  = 00000023  SegDs  = 0000002B
    SegSs  = 0000002B  SegEs  = 0000002B
    SegFs  = 00000053  SegGs  = 0000002B
    Dr0    = 00000000  Dr3    = 00000000
    Dr1    = 00000000  Dr6    = 00000000
    Dr2    = 00000000  Dr7    = 00000000
Call Stack (most recent call first):
  C:/projects/neovim/cmake/WindowsDllCopy.cmake:19 (get_prerequisites)
  
src\nvim\CMakeFiles\external_swiss.exe.dir\build.make:56: recipe for target 'src/nvim/CMakeFiles/external_swiss.exe' failed
mingw32-make[2]: *** [src/nvim/CMakeFiles/external_swiss.exe] Error 1
mingw32-make[2]: Leaving directory 'C:/projects/neovim/build'
CMakeFiles\Makefile2:394: recipe for target 'src/nvim/CMakeFiles/external_swiss.exe.dir/all' failed
mingw32-make[1]: *** [src/nvim/CMakeFiles/external_swiss.exe.dir/all] Error 2
```

equalsraf commented:
@justinmk probably no linking, but trying to determine the DLL dependencies https://cmake.org/cmake/help/v3.0/module/GetPrerequisites.html to determine what DLLs to include (https://github.com/neovim/neovim/blob/master/cmake/WindowsDllCopy.cmake).

 But for swiss.exe its probably safe to skip this.

PS: Using dumpbin is a bit strange though, since its building with MSYS.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -299,7 +299,7 @@ if(WIN32)
-  foreach(BIN win32yank.exe)
+  foreach(BIN cat.exe curl.exe diff.exe tee.exe tidy.exe win32yank.exe)
     unset(BIN_PATH CACHE)
     find_program(BIN_PATH ${BIN})
```
@equalsraf Why do we use `find_program()` here instead of just copying the file?

justinmk committed:
win/package: nvim-qt GUI

Closes #6126

justinmk committed:
win/CI: Cache dependencies.

justinmk committed:
win/package: Copy externals without analyzing.

These are just blobs that we jammed into the package. find_program() and
WindowsDllCopy.cmake do not make sense here, they search include paths
and try to determine DLL dependencies (GetPrerequisites).

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -326,6 +307,34 @@ if(WIN32)
+  add_custom_target(external_blobs
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/windows_runtime_deps/platforms
+
+    COMMAND ${CMAKE_COMMAND} -E copy "${DEPS_PREFIX}/bin/ca-bundle.crt"       ${PROJECT_BINARY_DIR}/windows_runtime_deps/
```
@equalsraf Here's what I'm doing instead of the `find_program()` + `WindowsDllCopy.cmake` approach (notice the deleted lines just above). It's a bit inelegant: because `find_program()` is not used, this is referencing `${DEPS_PREFIX}/bin/` which means this target has implicit knowledge about the `third-party/CMakeLists.txt` project.

But I don't know a better way. `WindowsDllCopy.cmake` was running `dumpbin` on the `.exe` files, which seems  useless (and wrong) since they have nothing to do with the Neovim build, we only want these `.exe` files for inclusion with the end-user package.

equalsraf replied:
> WindowsDllCopy.cmake was running dumpbin on the .exe files, which seems useless (and wrong

The dumpbin step is used to determine the DLLs that need to be shipped with each .exe i.e. part of the list you added below by hand. The advantage was that this would work with any binary in path.

swiss.exe seems to be self contained (does it have any DLL dependencies?). win32yank does need a couple DLLs.

Concerning nvim-qt, if it is never used during the tests or build, then there is no need to put it in DEPS_INSTALL and mix it with the other files there, only to separate it later with a long list of copy commands. This is different from diff or win32yank that should be available for test runs. In fact you could run that GetBinaryDep install recipe right into the nvim target install dir (problem with that is third-party does not know the install path).  Since technically this is not a build dependency, maybe the recipe could just be in the main cmake project as a post-install target.

jmlucjav commented:
not sure if this is relevant, just read about it and mentioning it just in case: https://blogs.technet.microsoft.com/virtualization/2017/12/19/tar-and-curl-come-to-windows/

of course only available in Insider Build 17063 and newer...
(I commented on the wrong issue before)
